constraints_lang.md

```diff
@@ -0,0 +1,311 @@
+
+`show` can generally be omitted for simple scripts or individual parts. However, it's critical for libraries or larger assemblies where program modification can cause geometry to unintentionally become partially constrained. In these cases, the user would want to be notified instead of geometry accidentally disappearing.
+
+### Solver Hints
+
+When users create complex relations, it may be outside of the ability of the implementation to solve for unknowns. Instead of restricting the types of relations that users can have, we give the user the ability to enter additional equations to allow fully constraining geometry.
+
+For example, say the user wants to do this:
+
+```
+derivative(f, x) = 0
+```
+
+Depending on the abilities of the solver, it may not be able to meaningfully use this to constrain `x`. Similar to `unsafe` in Rust, `as` in TypeScript, or other languages where you can tell the system, "Trust me. I've verified this myself, and you can use it."
+
+```
+manually solve derivative(f, x) = 0 using {
+  x = 2 * y - 1
+}
+```
+
+The above is intentionally verbose and cumbersome in order to discourage its use. Similar to `unsafe` in Rust or `as` in TypeScript that disables certain checks that the language usually does, this tells the solver that when the first constraint is applied, it can optionally ignore the opaque constraint and use the additional constraints instead.
+# Implementation
+
+## Language
+
+```
+statement ::= r | decl | show e | hide e
+relation ::= e | r rel_op r | minimize(e) | maximize(e)
+relation_operator ::= `=` | `<` | `>` | `<=` | `>=` | `or`
+declaration ::=
+ | fn f(x0, ..., xn) { s0; ...; sn }
+ | constraint c(x0, ..., xn) { s0, ..., sn }
+expression ::=
+ | n
+ | un_op e
+ | e bin_op e
+ | x
+ | e(e0, ...., en)
+ | [e0, ..., en]
+ | e[e]
+ | { l0: e, ..., ln: e}
+ | e.l
+ | `manually solve` c0; ...; cn `using` { c0; ...; cn }
+```
+
+In terms of the language definition, resolution strategies are special built-in relations.
+
+Negation is omitted from relation operators since this complicates the implementation.
+
+Let-bindings for defining new variables are not statements. Instead, it's a relation that creates an equality constraint. There's no difference between defining a variable and creating an equality constraint.
+
+Relations can use pure functions to constrain, and pure functions can use relations to constrain a value and return it. But pure functions cannot be used to constrain their arguments.
+## Execution
+
+As a user, you can think of execution as building up constraints, symbolic manipulation to solve for unknowns, resolution strategies, and numeric computation, in that order. Concrete numbers are only used in the last step.
+
+However, the way this is actually implemented is by mapping language constructs to Datalog:
```

Comment:
Am I understanding correctly that this would actually *use* Datalog under the hood, or just conceptually, and continue to be in Rust?

Reply:
I imagine that we'd either write our own Datalog engine in Rust or use a Rust library, like [this one](https://github.com/rust-lang/datafrog). The Rust compiler uses [chalk](https://github.com/rust-lang/chalk), which is another guide.